### ISSUE TITLE
Time comparison with pytorch communication hook impl

### ISSUE THREAD

**Issue by EtoDemerzel0427** [epfml/powersgd#18]
Hi, the time per batch experiments has a Runtime Error and thus are not available: https://observablehq.com/@tvogels/powersgd-benchmark. This paper(https://arxiv.org/pdf/2103.00543.pdf) stated that the communication hook would cause some overhead because of the unwanted overlap, so I really want to have a look at your benchmark results to check with their findings. Thanks.

**Comment by tvogels** [comment]
Hi, thanks for letting us know! I've fixed the timing results on that blog post now. I am not sure if those results are detailed enough to answer your question, unfortunately.

I've read the paper you shared, and their finding that PowerSGD is faster without overlapping communication and gradient computation than with is very interesting. I cannot explain why that would be the case, and if it's an implementation detail or something more fundamental.

If you have any questions regarding efficient use of PowerSGD, please do not hesitate to reach out.

**Comment by EtoDemerzel0427** [comment]
Thank you for your immediate reply and fix. From your experiments on cifar-10 it looks like these two implementations have comparable training speeds. But yes since the experiment settings were limited, it's not sufficient to draw a conclusion or something.

The authors of that paper said that since gradient compression methods were actually doing heavy computations, so the overlapping would be between "compression computation" and "backward pass computation", both of which use the GPU resources so they are not so good in parallelism. I'd say this argument sounds valid to me, though it's a bit unfortunate for me to find this paper only after I've implemented my own communication hook lol.

Anyway, thanks for your help!

**Comment by tvogels** [comment]
> The authors of that paper said that since gradient compression methods were actually doing heavy computations, so the overlapping would be between "compression computation" and "backward pass computation", both of which use the GPU resources so they are not so good in parallelism. I'd say this argument sounds valid to me, though it's a bit unfortunate for me to find this paper only after I've implemented my own communication hook lol.

While it makes sense to me that the gradient computation cannot be perfectly overlapped with PowerSGD computations, for me the surprising observation is that it seems to take *longer* than doing the gradient computation first, and then the same PowerSGD computations only afterwards (as we did it in the paper). This seems to hint at GPU scheduling difficulties rather than a problem with the actual execution of the operations.

Did you observe the same behavior with your own hook?

**Comment by EtoDemerzel0427** [comment]
@tvogels I am still doing some very basic experiments on my own hook, and I haven't implemented a distributed optimizer version, so currently, I don't have such results to validate it. 

About their observation, I agree with you that this is likely to be a GPU scheduling problem.  Compared with the distributed optimizer implementation, In the communication hook, the gradient computation and powersgd computation will compete for GPU resources, thus this part could have some overhead due to the scheduling (in my understanding, if they are sequentially executed, the GPU's access pattern will be more regular so that can maximize GPU's power); the communication primitive is still asynchronous, and thus has the potential to bring efficiency. But for gradient compression methods, the amount of communication is reduced, so this part won't be able to cover the overhead. So logically this makes sense to me.

If I have time I would be happy to profile it myself some time to check with this argument experimentally.
